src/kernel/include/pgtbl.h

```diff
@@ -57,245 +70,398 @@ struct tlb_quiescence {
+#define COS_PGTBL_SIZE_2K      11
+#define COS_PGTBL_SIZE_4K      12
+#define COS_PGTBL_SIZE_8K      13
+#define COS_PGTBL_SIZE_16K     14
+#define COS_PGTBL_SIZE_32K     15
+#define COS_PGTBL_SIZE_64K     16
+#define COS_PGTBL_SIZE_128K    17
+#define COS_PGTBL_SIZE_256K    18
+#define COS_PGTBL_SIZE_512K    19
+#define COS_PGTBL_SIZE_1M      20
+#define COS_PGTBL_SIZE_2M      21
+#define COS_PGTBL_SIZE_4M      22
+
+#define COS_PGTBL_NUM_2        1
+#define COS_PGTBL_NUM_4        2
+#define COS_PGTBL_NUM_8        3
+#define COS_PGTBL_NUM_16       4
 
-/* make it an opaque type...not to be touched */
-typedef struct pgtbl * pgtbl_t;
-
-/* identical to the capability structure */
-struct cap_pgtbl {
-	struct cap_header h;
-	u32_t refcnt_flags;          /* includes refcnt and flags */
-	pgtbl_t pgtbl;
-	u32_t lvl; 		     /* what level are the pgtbl nodes at? */
-	struct cap_pgtbl *parent;    /* if !null, points to parent cap */
-	u64_t frozen_ts;             /* timestamp when frozen is set. */
-} __attribute__((packed));
-
-static pgtbl_t pgtbl_alloc(void *page)
-{ return (pgtbl_t)((unsigned long)__pgtbl_alloc(&page) & PGTBL_FRAME_MASK); }
-
-static void
-pgtbl_init_pte(void *pte)
+/*
+union pgtbl_data
 {
-	int i;
-	unsigned long *vals = pte;
-
-	for (i = 0 ; i < (1<<PGTBL_ORD) ; i++) vals[i] = 0;
-}
-
-static int
-pgtbl_intern_expand(pgtbl_t pt, u32_t addr, void *pte, u32_t flags)
+	struct MPU;
+	struct pgtbl;
+};
+*/
+struct pgtbl  /* 9 words */
 {
-	unsigned long accum = (unsigned long)flags;
-	int ret;
-
-	/* NOTE: flags currently ignored. */
-
-	assert(pt);
-	assert((PGTBL_FLAG_MASK & (u32_t)pte) == 0);
-	assert((PGTBL_FRAME_MASK & flags) == 0);
+	u32_t type_addr;
+	/* For MPU there are 5(4) regions' space, for pgtbl there are 8 pages' space. The start address is stored in the child nodes.
+	 * We no longer allocate memory for page tables. we always use 8 pages. data[9] will contain the top-level addr, if needed */
+	u32_t data[10];
+};
 
-	if (!pte) return -EINVAL;
-	ret = __pgtbl_expandn(pt, (unsigned long)(addr >> PGTBL_PAGEIDX_SHIFT),
-			      PGTBL_DEPTH, &accum, &pte, NULL);
-	if (!ret && pte) return -EEXIST; /* no need to expand */
-	assert(!(ret && !pte));		 /* error and used memory??? */
+/* identical to the capability structure */
+struct cap_pgtbl {     /* 16 words */
+	struct cap_header h;                                              /* 1 */
+	u32_t refcnt_flags;          /* includes refcnt and flags */      /* 1 */
+	struct pgtbl pgtbl;                                               /* 11 */
+	struct cap_pgtbl *parent;    /* if !null, points to parent cap */ /* 1 */
+	u64_t frozen_ts;             /* timestamp when frozen is set. */  /* 2 */
+} __attribute__((packed));
 
-	return ret;
-}
+/* make it an opaque type...not to be touched */
+/* PRY:We change pgtbl_t here. How do we inline these ? */
+/* inline the MPU representation into the pgtbl caps:MPU needs, say, 4 slots, which is 8 words */
+typedef struct pgtbl * pgtbl_t;
```

Comment:
yeah, looks like you can't really use this anywhere, right?  How can you hide the representation?  How much of the existing code needs to be changed for this?